blog/2024-08-12-ts-api-refactor.md

```diff
@@ -0,0 +1,449 @@
+---
+slug: ts-api-refactor
+title: Typescript API Package Auth Refactor
+tags: []
+---
+
+Today we are merging some changes to how the TypeScript `@atproto/api` package works with authentication sessions. The changes are mostly backwards compatible, but some parts are now deprecated, and there are some breaking changes for advanced uses.
+
+The motivation for these changes is the need to make the `@atproto/api` package compatible with OAuth session management. We don't have OAuth client support "launched" and documented quiet yet, so you can keep using the current app password authentication system. When we do "launch" OAuth support and start encourage it's usage in the near future (see the [OAuth Roadmap](https://github.com/bluesky-social/atproto/discussions/2656)), these changes will make it easier to migrate.
+
+In addition, the redesigned session management system fixes a bug that could cause the session data to become invalid when Agent clones are created (e.g. using `agent.withProxy()`).
+
+## New Features
+
+We've restructured the `XrpcClient` HTTP fetch handler to be specified during the instantiation of the XRPC client, through the constructor, instead of using a default implementation (which was statically defined).
+
+With this refactor, the XRPC client is now more modular and reusable. Session management, retries, cryptographic signing, and other request-specific logic can be implemented in the fetch handler itself rather than by the calling code.
+
+A new abstract class named `Agent`, has been added to `@atproto/api`. This class will be the base class for all Bluesky agents classes in the `@atproto` ecosystem. It is meant to be extended by implementation that provide session management and fetch handling for the `AtpClient` instances.
```

Comment:
```suggestion
A new abstract class named `Agent`, has been added to `@atproto/api`. This class will be the base class for all Bluesky agents classes in the `@atproto` ecosystem. It is meant to be extended by implementation that provide session management and fetch handling.
```

Not sure what I meant by that. `AtpClient` is kind of an internal implementation detail. I updated the client code & changeset accordingly.